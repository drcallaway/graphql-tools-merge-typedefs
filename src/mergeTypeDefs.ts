import { ITypeDefinitions } from 'graphql-tools';

/**
 * Merges all definitions of the given operation type.
 *
 * @param defs Type definitions to merge
 * @param type Operation type to merge
 */
function merge(defs: ITypeDefinitions, type: String): ITypeDefinitions {
  let result;

  defs = (defs as any).filter(def => {
    // filter out falsy values just incase
    if (!def) return false;

    // reset def.definitions to be the filtered version so the `type` that
    // was passed in doesn't exist in it anymore
    def.definitions = def.definitions.filter(definition => {
      const { kind, name, fields } = definition;
      if (kind === 'ObjectTypeDefinition' && name.kind === 'Name' && name.value === type) {
        if (!result) {
          result = definition;
        } else {
          result.fields = result.fields.concat(fields);
        }

        // remove the definition because it's been added to the result and
        // shouldn't exist here anymore
        return false;
      }

      return true;
    });

    if (def.definitions.length === 0) {
      return false; // operation was the only definition so this can be removed
    }

    return true;
  });

  // update to be `result?.fields?.length` when it's available
  if (((result || {}).fields || []).length) {
    if (result.fields.length > 1) {
      // filters out duplicate keys. The last key always wins
      result.fields = result.fields.filter((field, i, array) => {
        return !array.slice(i + 1).find((obj) => obj.name.value === field.name.value);
      });
    }

    (defs as any).push({
      kind: "Document",
      definitions: [result]
    });
  }

  return defs;
}

/**
 * Merges all query, mutation, and subscription type definitions.
 *
 * @param defs Type definitions to merge
 * @param types any additonal types you want to target
 */
function mergeTypeDefs(defs, types = []): ITypeDefinitions {
  // handles strings and arrays of strings
  // also always includes these base 3
  types = [ 'Query', 'Mutation', 'Subscription'].concat(types);
  return types.reduce((prev, next) => merge(prev, next), defs);
}

// export default as well so people can rename it to be what ever they want
export default mergeTypeDefs;
export { mergeTypeDefs, merge };
