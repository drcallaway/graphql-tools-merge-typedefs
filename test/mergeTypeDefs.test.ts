import gql from 'graphql-tag'
import { snapshot } from './utils';
import { mergeTypeDefs, merge } from '../src/mergeTypeDefs';


test.each([ 'Query', 'Mutation', 'Subscription' ])('%s type defs are merged', (type) => {
  const defs: any = mergeTypeDefs([
    gql`
      type ${type} {
        ${type.toLowerCase()}Field1: String
      }
    `,
    gql`
      type ${type} {
        ${type.toLowerCase()}Field2: String
      }
    `,
  ])

  snapshot(defs)
})


test('all type defs are merged', () => {
  const defs: any = mergeTypeDefs([
    gql`
      type Query {
        queryFieldSet1: String
      }
      type Mutation {
        mutationFieldSet1: String
      }
      type Subscription {
        subscriptionFieldSet1: String
      }
    `,
    gql`
      type Query {
        queryFieldSet2: String
      }
      type Mutation {
        mutationFieldSet2: String
      }
      type Subscription {
        subscriptionFieldSet2: String
      }
    `,
  ])

  snapshot(defs)
});

test('two fields in the same type', () => {
  const defs: any = mergeTypeDefs([
    gql`
      type Local {
        field1: String
        field2: String
      }
    `,
    gql`
      type Local {
        field1: Int
      }
    `,
  ], 'Local')

  snapshot(defs)
})


test('multiple types in a def', () => {
  const defs: any = mergeTypeDefs([
    gql`
      type Fun {
        woohoo: String
      }
      type User {
        id: String
      }
      type Local {
        field1: User
        field2: Fun
      }
    `,
    gql`
      type Local {
        field3: Int
      }
    `,
  ], 'Local')

  // make sure the type is removed from each def except the last one
  expect(defs.slice(0, -1).find((def) => def.definitions.find(({ name }) => name.value === 'Local'))).toBeFalsy()
  snapshot(defs)
})


test('filters out falsy values', () => {
  const defs: any = mergeTypeDefs([
    false,
    null,
    gql`
      type Something {
        field1: Int
      }
    `,
    gql`
      type Query {
        something: Something
      }
    `,
  ], 'Something')

  snapshot(defs)
})

test('handles multiple types', () => {
  const defs: any = mergeTypeDefs([
    gql`
      type Local {
        field1: Int
      }
      type Something {
        field1: Int
      }
    `,
    gql`
      type Local {
        field2: Int
      }
      type Something {
        field2: Int
      }
      type Query {
        something: Something
      }
    `,
  ], [ 'Something', 'Local' ])

  snapshot(defs)
})
