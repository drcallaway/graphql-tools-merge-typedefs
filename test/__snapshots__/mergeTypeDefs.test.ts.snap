// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation type defs are merged 1`] = `
"type Mutation {
  mutationField1: String
  mutationField2: String
}
"
`;

exports[`Query type defs are merged 1`] = `
"type Query {
  queryField1: String
  queryField2: String
}
"
`;

exports[`Subscription type defs are merged 1`] = `
"type Subscription {
  subscriptionField1: String
  subscriptionField2: String
}
"
`;

exports[`all type defs are merged 1`] = `
"type Query {
  queryFieldSet1: String
  queryFieldSet2: String
}

type Mutation {
  mutationFieldSet1: String
  mutationFieldSet2: String
}

type Subscription {
  subscriptionFieldSet1: String
  subscriptionFieldSet2: String
}
"
`;

exports[`filters out falsy values 1`] = `
"type Query {
  something: Something
}

type Something {
  field1: Int
}
"
`;

exports[`multiple types in a def 1`] = `
"type Fun {
  woohoo: String
}

type User {
  id: String
}

type Local {
  field1: User
  field2: Fun
  field3: Int
}
"
`;

exports[`two fields in the same type 1`] = `
"type Local {
  field2: String
  field1: Int
}
"
`;
